ProjFS.Mac/PrjFSKextTests/VnodeCacheTests.mm

```diff
@@ -0,0 +1,174 @@
+#import <XCTest/XCTest.h>
+
+typedef int16_t VirtualizationRootHandle;
+
+#include "../PrjFSKext/VnodeCacheTestable.hpp"
+
+@interface VnodeCacheTests : XCTestCase
+@end
+
+@implementation VnodeCacheTests
+
+struct VnodeCacheEntry
+{
+    vnode_t vnode;
+    uint32_t vid;   // vnode generation number
+    VirtualizationRootHandle virtualizationRoot;
+};
+
+static const VirtualizationRootHandle TestVirtualizationRootHandle = 3;
+
+VirtualizationRootHandle VirtualizationRoot_FindForVnode(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter functionCounter,
+    PrjFSPerfCounter innerLoopCounter,
+    vnode_t _Nonnull vnode,
+    const FsidInode& vnodeFsidInode);
+
+VirtualizationRootHandle VirtualizationRoot_FindForVnode(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter functionCounter,
+    PrjFSPerfCounter innerLoopCounter,
+    vnode_t _Nonnull vnode,
+    const FsidInode& vnodeFsidInode)
+{
+    return TestVirtualizationRootHandle;
+}
+
+- (void)testHashVnodeWithCapacityOfOne {
+    s_entriesCapacity = 1;
+    XCTAssertTrue(0 == HashVnode(reinterpret_cast<vnode_t>(static_cast<uintptr_t>(1))));
+}
+
+- (void)testInvalidateCache_ExclusiveLocked_SetsMemoryToZeros {
+    s_entriesCapacity = 100;
+    VnodeCacheEntry* emptyArray = static_cast<VnodeCacheEntry*>(calloc(s_entriesCapacity, sizeof(VnodeCacheEntry)));
+    s_entries = static_cast<VnodeCacheEntry*>(calloc(s_entriesCapacity, sizeof(VnodeCacheEntry)));
+    
+    memset(s_entries, 1, s_entriesCapacity * sizeof(VnodeCacheEntry));
+    XCTAssertTrue(0 != memcmp(emptyArray, s_entries, sizeof(VnodeCacheEntry) * s_entriesCapacity));
+    
+    InvalidateCache_ExclusiveLocked();
+    XCTAssertTrue(0 == memcmp(emptyArray, s_entries, sizeof(VnodeCacheEntry)*s_entriesCapacity));
+    
+    free(s_entries);
+    free(emptyArray);
+}
+
+- (void)testTryFindVnodeIndex_SharedLocked_ReturnsStartingIndexWhenNull {
+    s_entriesCapacity = 100;
+    s_entries = static_cast<VnodeCacheEntry*>(calloc(s_entriesCapacity, sizeof(VnodeCacheEntry)));
+    
+    vnode_t testVnode = reinterpret_cast<vnode_t>(static_cast<uintptr_t>(1));
```

Comment:
I suspect this is technically undefined behaviour. How about declaring a `static` dummy vnode?